### PR TITLE
feat: add option to filter licenses list

### PR DESCRIPTION
**Pull request by aiji42**
closes https://github.com/pnpm/pnpm/issues/5806

## Overview

This PR introduces an option to output licenses specifically for selected workspaces when executing `pnpm licenses list`.

## Details

The `--filter` option has been added, which provides the ability to filter license results.

Example usage:
```
❯ pnpm licenses list -h  
Version 8.9.0
Usage: pnpm licenses ls
       pnpm licenses list
       pnpm licenses list --long

Check the licenses of the installed packages.

Options:
  -D, --dev                Check only "devDependencies"
      --json               Show information in JSON format
      --long               Show more details (such as a link to the repo) are not displayed. To display the details, pass this option.
      --no-optional        Don't check "optionalDependencies"
  -P, --prod               Check only "dependencies" and "optionalDependencies"

Filtering options (run the command only on packages that satisfy at least one of the selectors):
      --changed-files-ignore-pattern <pattern>  Defines files to ignore when filtering for changed projects since the specified commit/branch. Usage example: pnpm --filter="...[origin/master]" --changed-files-ignore-pattern="**/README.md" build
      --filter !<selector>                      If a selector starts with ! (or \! in zsh), it means the packages matching the selector must be excluded. E.g., "pnpm --filter !foo" selects all packages except "foo"
      --filter .                                Includes all packages that are under the current working directory
      --filter ...^<pattern>                    Includes only the direct and indirect dependents of the matched packages without including the matched packages themselves. ^ must be doubled at the Windows Command Prompt. E.g.: ...^foo (...^^foo in Command Prompt)
      --filter ...<pattern>                     Includes all direct and indirect dependents of the matched packages. E.g.: ...foo, "...@bar/*"
      --filter ./<dir>                          Includes all packages that are inside a given subdirectory. E.g.: ./components
      --filter "[<since>]"                      Includes all packages changed since the specified commit/branch. E.g.: "[master]", "[HEAD~2]". It may be used together with "...". So, for instance, "...[HEAD~1]" selects all packages changed in the last commit and their dependents
      --filter {<dir>}                          Includes all projects that are under the specified directory. It may be used with "..." to select dependents/dependencies as well. It also may be combined with "[<since>]". For instance, all changed projects inside a directory:
                                                "{packages}[origin/master]"
      --filter <pattern>                        Restricts the scope to package names matching the given pattern. E.g.: foo, "@bar/*"
      --filter <pattern>...                     Includes all direct and indirect dependencies of the matched packages. E.g.: foo...
      --filter <pattern>^...                    Includes only the direct and indirect dependencies of the matched packages without including the matched packages themselves. ^ must be doubled at the Windows Command Prompt. E.g.: foo^... (foo^^... in Command Prompt)
      --filter-prod <pattern>                   Restricts the scope to package names matching the given pattern similar to --filter, but it ignores devDependencies when searching for dependencies and dependents.
      --test-pattern <pattern>                  Defines files related to tests. Useful with the changed since filter. When selecting only changed packages and their dependent packages, the dependent packages will be ignored in case a package has changes only in tests. Usage example: pnpm
                                                --filter="...[origin/master]" --test-pattern="test/*" test

Visit https://pnpm.io/8.x/cli/licenses for documentation about this command.
```

Standard output without using the filter:
```
> pnpm licenses list -P
┌────────────────────────────────────────┬───────────────────────────┐
│ Package                                │ License                   │
├────────────────────────────────────────┼───────────────────────────┤
│ json-schema                            │ (AFL-2.1 OR BSD-3-Clause) │
├────────────────────────────────────────┼───────────────────────────┤
│ JSONStream                             │ (MIT OR Apache-2.0)       │
├────────────────────────────────────────┼───────────────────────────┤
│ type-fest                              │ (MIT OR CC0-1.0)          │
├────────────────────────────────────────┼───────────────────────────┤
│ tslib                                  │ 0BSD                      │
so match more ...
```

Output when using the `--filter` option:
```
❯ pnpm licenses list --filter matcher -P           
┌──────────────────────┬─────────┐
│ Package              │ License │
├──────────────────────┼─────────┤
│ escape-string-regexp │ MIT     │
└──────────────────────┴─────────┘
```

### CODE REVIEW

**Review by zkochan — CHANGES_REQUESTED** [review]
We already have a `--filter` flag and `selectedProjectsGraph` is passed down to all plugins. The already present filtering mechanism should be used.

**Comment by aiji42** [comment]
@zkochan I fixed to use `--filter` 🙆‍♂️

File: reviewing/license-scanner/src/licenses.ts [hunk 1]
```diff
@@ -73,6 +73,7 @@ export async function findDependencyLicenses (opts: {
   modulesDir?: string
   registries: Registries
   wantedLockfile: Lockfile | null
+  importerIds?: string[]
```
Comment (zkochan) [inline reviewing/license-scanner/src/licenses.ts]: let's call it
```suggestion
  includedImporterIds?: string[]
```

File: reviewing/plugin-commands-licenses/test/fixtures/workspace-licenses/bar/package.json [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  "name": "bar",
+  "version": "1.0.0",
+  "description": "",
+  "main": "index.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "keywords": [],
+  "author": "",
+  "license": "ISC",
```
Comment (zkochan) [inline reviewing/plugin-commands-licenses/test/fixtures/workspace-licenses/bar/package.json]: remove these unnecessary fields

File: reviewing/plugin-commands-licenses/test/index.ts [hunk 3]
```diff
@@ -170,27 +171,43 @@ test('pnpm licenses should work with file protocol dependency', async () => {
   expect(stripAnsi(output)).toMatchSnapshot('show-packages')
 })
 
-test('pnpm licenses should work with git protocol dep that have patches', async () => {
+test('pnpm licenses: filter outputs', async () => {
```
Comment (zkochan) [inline reviewing/plugin-commands-licenses/test/index.ts]: Why was an existing test removed?

**Review by zkochan — CHANGES_REQUESTED** [review]

File: reviewing/plugin-commands-licenses/test/index.ts (hunk 3 above)
Comment (aiji42) [reply reviewing/plugin-commands-licenses/test/index.ts]: It was a mistake. I've corrected it.

**Comment by aiji42** [comment]
@zkochan Made the changes. Thank you for the feedback! Please review again.

**Review by zkochan — APPROVED** [review]

**Pull request merged**